src/main/java/org/broadinstitute/hellbender/utils/nio/NioBam.java

```diff
@@ -0,0 +1,161 @@
+package org.broadinstitute.hellbender.utils.nio;
+
+import com.google.cloud.storage.contrib.nio.CloudStorageFileSystem;
+import htsjdk.samtools.BAMFileSpan;
+import htsjdk.samtools.BAMIndex;
+import htsjdk.samtools.Chunk;
+import htsjdk.samtools.QueryInterval;
+import htsjdk.samtools.SAMFileHeader;
+import htsjdk.samtools.SAMRecord;
+import htsjdk.samtools.SAMSequenceRecord;
+import htsjdk.samtools.SamInputResource;
+import htsjdk.samtools.SamReader;
+import htsjdk.samtools.SamReaderFactory;
+import htsjdk.samtools.ValidationStringency;
+import htsjdk.samtools.seekablestream.ByteArraySeekableStream;
+import htsjdk.samtools.seekablestream.SeekableStream;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.broadinstitute.hellbender.utils.gcs.BucketUtils;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.Serializable;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * NioBam holds a reference to a BAM file on Google Cloud Storage, and can give you
+ * an RDD with the reads inside of it.
+ *
+ * The constructor will open the file to make sure it exists and we have access to it.
+ * This is preferable to waiting until we're running on the cloud.
+ *
+ * Although we don't expect you to move these objects across computers, you can.
+ */
+public class NioBam implements Serializable {
+
+    private static final long serialVersionUID = 1L;
+    final String bam;
+    final String index;
+
+    transient byte[] indexCache;
+
+    /** Checks the files exists, then stores them. **/
+    public NioBam(String gcsFilename, String indexGcsFilename) throws IOException {
+        this.bam = gcsFilename;
+        this.index = indexGcsFilename;
+        Path bamPath = BucketUtils.getPathOnGcs(bam);
+        Path bamIndexPath = BucketUtils.getPathOnGcs(index);
+        if (!Files.exists(bamPath)) {
```

Comment:
About 200-300ms. The point is that it's a lot less than the latency of starting a Spark job only to have it fail because of a typo in a filename.

This may move somewhere else (especially since I understand there's work on a path class of our own) but I think that the key idea of having a class to hold paths, with the invariant that you can only create one of those if the file exists, is worth keeping. It also means that we do the check only once, since everyone who receives such an object knows the path has been checked already.